lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp

```diff
@@ -508,6 +511,319 @@ tryEliminateUnneededForwardingInst(SILInstruction *i,
   return None;
 }
 
+//===----------------------------------------------------------------------===//
+//                   DestructureTuple <-> Tuple Round Trip
+//===----------------------------------------------------------------------===//
+
+/// hat(%x) = destructure_tuple x
+/// %x2 = tuple(hat(%x))
+/// ->
+/// %x2 = %x
+static SILBasicBlock::iterator
```

Comment:
Not really the point of this PR, but I suggest to do it now: Can you use the `InstructionDeleter` in `CanonicalizeInstruction` to get rid of this iterator updating/passing/returning madness?